src/Modules/UngradedModules.jl

```diff
@@ -2792,61 +3259,172 @@ end
 # dual: ambig: hom(M, R) or hom(M, Q(R))?
 
-@doc Markdown.doc"""
-    coordinates(a::FreeModElem, SQ::SubQuo)::Union{Nothing,Oscar.SRow}
+function lift_with_unit(a::FreeModElem{T}, generators::ModuleGens{T})::Union{Nothing,Oscar.SRow} where {T <: MPolyElem}
+  # TODO allow optional argument ordering
+  # To do this efficiently we need better infrastructure in Singular.jl
+  R = base_ring(parent(a))
+  singular_assure(generators)
+  if Singular.has_global_ordering(base_ring(generators.SF))
+    l = lift(a, generators)
+    return l, R(1)
+  end
+  error("Not implemented")
+end
 
-Compute a sparse row `r` such that `a` is a representative of `SubQuoElem(r, SQ)`.
-If no such `r` exists, `nothing` is returned.
-"""
-function coordinates(a::FreeModElem, M::SubModuleOfFreeModule)::Union{Nothing,Oscar.SRow}
+function lift(a::FreeModElem{T}, generators::ModuleGens{T})::Union{Nothing,Oscar.SRow} where {T <: MPolyElem}
   if iszero(a)
     return sparse_row(base_ring(parent(a)))
   end
-  singular_assure(M.gens)
-  S = singular_generators(M.gens)
-  b = ModuleGens([a], M.gens.SF)
+  singular_assure(generators)
+  S = singular_generators(generators)
+  b = ModuleGens([a], generators.SF)
   singular_assure(b)
   s, r = Singular.lift(S, singular_generators(b))
   if Singular.ngens(s) == 0 || iszero(s[1])
     return nothing
   end
-  Rx = base_ring(M)
-  return sparse_row(Rx, s[1], 1:ngens(M))
+  Rx = base_ring(generators)
+  return sparse_row(Rx, s[1], 1:ngens(generators))
 end
 
 @doc Markdown.doc"""
-    coordinates(a::FreeModElem, SQ::SubQuo)::Union{Nothing,Oscar.SRow}
-
-Compute a sparse row `r` such that `a` is a representative of `SubQuoElem(r, SQ)`.
+    coordinates(a::FreeModElem{T}, generators::ModuleGens{T})::Union{Nothing,Oscar.SRow}
+Compute a sparse row `r` such that `a = sum([r[i]*gen(generators,i) for i in 1:ngens(generators)])`.
 If no such `r` exists, `nothing` is returned.
 """
-function coordinates(a::FreeModElem, SQ::SubQuo)::Union{Nothing,Oscar.SRow}
-  coords = coordinates(a, SQ.sum)
-  if coords !== nothing
-    return return coords[1:ngens(SQ)]
-  else
-    return coords
+function coordinates(a::FreeModElem{T}, generators::ModuleGens{T})::Union{Nothing,Oscar.SRow} where {T <: MPolyElem}
+  singular_assure(generators)
+  if !Singular.has_global_ordering(base_ring(generators.SF))
+    error("Ordering must be global")
   end
+  return lift(a, generators)
+end
+
+@doc Markdown.doc"""
+    coordinates_via_transform(a::FreeModElem{T}, generators::ModuleGens{T})::Union{Nothing,Oscar.SRow} where T
+Let `generators` be a Gröbner basis and let `A*M = generators`.
+Compute a sparse row `r` such that `a = sum([r[i]*gen(M,i) for i in 1:ngens(M)])`.
+If no such `r` exists, `nothing` is returned.
+"""
+function coordinates_via_transform(a::FreeModElem{T}, generators::ModuleGens{T})::Union{Nothing,Oscar.SRow} where T
+  A = get_attribute(generators, :transformation_matrix)
+  a === nothing && error("No transformation matrix in the Gröbner basis.")
```

Comment:
Don't you mean this???
```suggestion
  A === nothing && error("No transformation matrix in the Gröbner basis.")
```